### ISSUE TITLE
Function parameters that implement `Copy` should be passed by value

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#2153]
We should use pass by value for parameters that implement `Copy` i.e., signatures, keys, hashes etc. 

For example `Witness::p2wpkh`.

This was recently discussed, I can't remember where right now, link to discussion or Rust documentation citation needed.

**Comment by tcharding** [comment]
Ouch, clearly didn't search for that did I, its in the name :( I definitely owe you a drink for that.

**Comment by sanket1729** [comment]
I agree with this for all rust-bitcoin/rust-secp types, but we should be careful with this in general when dealing with large values. See the clippy pedantic lint [here](https://rust-lang.github.io/rust-clippy/master/index.html#large_types_passed_by_value)

**Comment by tcharding** [comment]
> everything that is `Copy` and 32 bytes or less

I went to edit the title and description and relized that this _excludes_ signatures. We have at least one place in the codebase that violates this

```
    pub fn from_signature(sig: Signature) -> SerializedSignature { sig.serialize() }
```

So this issue is now two issues:

1. Pass by value for everything that is `Copy` and less <= 32 bytes
2. Pass by reference for everything that is > 32 bytes

**Comment by apoelstra** [comment]
Ok, let's make the threshold 64 bytes then :). This will also cover the SHA-512 hashes used in bip32.

**Comment by apoelstra** [comment]
Oh, I guess signatures are still (sometimes) 65+ bytes ... ok, I think we should still pass them by value, under the "things larger than X bytes we will handle in an ad-hoc way" clause.

**Comment by sanket1729** [comment]
The clippy lint has a default value of 256. Seems like a good starting point.

**Comment by tcharding** [comment]
+1 to using the default lint value.
